hail/src/main/scala/is/hail/expr/ir/PruneDeadFields.scala

```diff
@@ -365,48 +397,71 @@ object PruneDeadFields {
+          key = childReqKey,
+          rowType = unify(child.typ.rowType, selectKey(child.typ.rowType, childReqKey), requestedType.rowType),
           globalType = requestedType.globalType), memo)
       case TableOrderBy(child, sortFields) =>
-        memoizeTableIR(child, child.typ.copy(
+        val k = if (sortFields.forall(_.sortOrder == Ascending) && child.typ.key.startsWith(sortFields.map(_.field)))
+          child.typ.key
+        else
+          FastIndexedSeq()
+        memoizeTableIR(child, TableType(
+          key = k,
           rowType = unify(child.typ.rowType,
-            child.typ.rowType.filterSet((sortFields.map(_.field) ++ child.typ.key).toSet)._1,
+            selectKey(child.typ.rowType, sortFields.map(_.field) ++ k),
             requestedType.rowType),
           globalType = requestedType.globalType), memo)
       case TableDistinct(child) =>
-        memoizeTableIR(child, requestedType, memo)
+        val dep = TableType(key = child.typ.key,
+          rowType = unify(child.typ.rowType, requestedType.rowType, selectKey(child.typ.rowType, child.typ.key)),
+          globalType = requestedType.globalType)
+        memoizeTableIR(child, dep, memo)
       case TableMapRows(child, newRow) =>
         val rowDep = memoizeAndGetDep(newRow, requestedType.rowType, child.typ, memo)
-        memoizeTableIR(child, unify(child.typ, minimal(child.typ).copy(globalType = requestedType.globalType), rowDep), memo)
+        val dep = TableType(
+          key = requestedType.key,
+          rowType = unify(child.typ.rowType, selectKey(requestedType.rowType, requestedType.key), rowDep.rowType),
+          globalType = unify(child.typ.globalType, requestedType.globalType, rowDep.globalType)
+        )
+        memoizeTableIR(child, dep, memo)
       case TableMapGlobals(child, newGlobals) =>
         val globalDep = memoizeAndGetDep(newGlobals, requestedType.globalType, child.typ, memo)
         memoizeTableIR(child, unify(child.typ, requestedType.copy(globalType = globalDep.globalType), globalDep), memo)
       case TableAggregateByKey(child, newRow) =>
         val aggDep = memoizeAndGetDep(newRow, requestedType.rowType, child.typ, memo)
-        memoizeTableIR(child, child.typ.copy(rowType = unify(child.typ.rowType, aggDep.rowType),
+        memoizeTableIR(child, TableType(key = child.typ.key,
+          rowType = unify(child.typ.rowType, aggDep.rowType, selectKey(child.typ.rowType, child.typ.key)),
           globalType = unify(child.typ.globalType, aggDep.globalType, requestedType.globalType)), memo)
       case TableKeyByAndAggregate(child, expr, newKey, _, _) =>
         val keyDep = memoizeAndGetDep(newKey, newKey.typ, child.typ, memo)
         val exprDep = memoizeAndGetDep(expr, requestedType.valueType, child.typ, memo)
         memoizeTableIR(child,
-          unify(child.typ, keyDep, exprDep, minimal(child.typ).copy(globalType = requestedType.globalType)),
+          TableType(
+            key = FastIndexedSeq(),
```

Comment:
Should this check if `newKey` is (a prefix of) `child.key`, or are we assuming the pruner always runs after a simplification pass?

Reply:
good question. I think we should leave this as is, but add a comment that if we start pruning before simplification, it would be possible to deoptimize.